recipes/mongo-c-driver/all/conanfile.py

```diff
@@ -0,0 +1,155 @@
+    }
+    generators = "cmake"
+    no_copy_source = True
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+        self._cmake.definitions["CMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP"] = "TRUE"
+        self._cmake.definitions["ENABLE_TESTS"] = "OFF"
+        self._cmake.definitions["ENABLE_EXAMPLES"] = "OFF"
+        # todo: apply patch https://github.com/microsoft/vcpkg/blob/master/ports/mongo-c-driver/disable-static-when-dynamic-build.patch
+        self._cmake.definitions["ENABLE_STATIC"] = "OFF" if self.options.shared else "ON"
+        self._cmake.definitions["ENABLE_SSL"] = self.options.ssl_provider
+        self._cmake.definitions["ENABLE_ZLIB"] = self.options.zlib
+        self._cmake.definitions["ENABLE_SHM_COUNTERS"] = "OFF"
+        if self.options.zlib == 'SYSTEM':
+            self._cmake.definitions["ZLIB_ROOT"] = self.deps_cpp_info["zlib"].rootpath
+        if self.options.ssl_provider == 'OPENSSL':
+            self._cmake.definitions["OPENSSL_ROOT_DIR"] = self.deps_cpp_info["openssl"].rootpath
+
+        self._cmake.configure(
+            source_folder=self._source_subfolder,
+            build_folder=self._build_subfolder
+        )
+
+        return self._cmake
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = "{name}-{version}".format(
+            name=self.name,
+            version=self.version
+        )
+        os.rename(extracted_dir, self._source_subfolder)
+        # todo: fix it https://github.com/bincrafters/community/issues/995
+        tools.replace_in_file(
+            os.path.join(self._source_subfolder, "src", "libmongoc", "CMakeLists.txt"),
+            "add_executable (mongoc-stat ${PROJECT_SOURCE_DIR}/../../src/tools/mongoc-stat.c)",
+            "# add_executable (mongoc-stat ${PROJECT_SOURCE_DIR}/../../src/tools/mongoc-stat.c)")
+        tools.replace_in_file(
+            os.path.join(self._source_subfolder, "src", "libmongoc", "CMakeLists.txt"),
+            "target_link_libraries (mongoc-stat mongoc_shared ${LIBRARIES})",
+            "# target_link_libraries (mongoc-stat mongoc_shared ${LIBRARIES})")
+        if self.options.ssl_provider == 'LIBRESSL':
+            tools.replace_in_file(
+                os.path.join(self._source_subfolder, "src", "libmongoc", "CMakeLists.txt"),
+                self._libressl_find_pattern,
+                self._libressl_replacement_pattern.format(
+                    LIBRESSL_INCLUDE_DIRS=self.deps_cpp_info["libressl"].include_paths[0],
+                    LIBRESSL_LIBRARY_DIRS=self.deps_cpp_info["libressl"].lib_paths[0]
+                ))
+
+    def build(self):
+        cmake = self._configure_cmake()
+        cmake.build()
```

Comment:
By moving the patches to build we ensure better functionality for different workflows

```suggestion
    def source(self):
        tools.get(**self.conan_data["sources"][self.version])
        extracted_dir = "{name}-{version}".format(name=self.name,version=self.version)
        os.rename(extracted_dir, self._source_subfolder)

    def _patch_sources(self):
        # todo: fix it https://github.com/bincrafters/community/issues/995
        tools.replace_in_file(
            os.path.join(self._source_subfolder, "src", "libmongoc", "CMakeLists.txt"),
            "add_executable (mongoc-stat ${PROJECT_SOURCE_DIR}/../../src/tools/mongoc-stat.c)",
            "# add_executable (mongoc-stat ${PROJECT_SOURCE_DIR}/../../src/tools/mongoc-stat.c)")
        tools.replace_in_file(
            os.path.join(self._source_subfolder, "src", "libmongoc", "CMakeLists.txt"),
            "target_link_libraries (mongoc-stat mongoc_shared ${LIBRARIES})",
            "# target_link_libraries (mongoc-stat mongoc_shared ${LIBRARIES})")
        if self.options.ssl_provider == 'LIBRESSL':
            tools.replace_in_file(
                os.path.join(self._source_subfolder, "src", "libmongoc", "CMakeLists.txt"),
                self._libressl_find_pattern,
                self._libressl_replacement_pattern.format(
                    LIBRESSL_INCLUDE_DIRS=self.deps_cpp_info["libressl"].include_paths[0],
                    LIBRESSL_LIBRARY_DIRS=self.deps_cpp_info["libressl"].lib_paths[0]
                ))

    def build(self):
        self._patch_sources()
        cmake = self._configure_cmake()
        cmake.build()
```

Reply:
These patches take a lot of space in the recipe and may change overtime, conan offers `tools.patch` to apply git diffs.

As a suggestion, it might be worth looking into here for this case. but not required at this point

Reply:
Move to a separate function